[QA-G1] OpenSearch Migration — Safe Startup When OpenSearch Is Unavailable (TC-001, TC-003–TC-005)

## Parent Epic

#35476

## Glossary (quick)

- **Old engine (ES)** — Elasticsearch, what dotCMS uses today. Port `9200`, inspected with **Kibana** (`5601`).
- **New engine (OS)** — OpenSearch, the migration target. Port `9201`, inspected with **OpenSearch Dashboards** (`5602`).
- **Migration phase** — `0` = old engine only · `1` = write to *both*, search the old one · `2` = write to both, read OS.
- **Dual-write** — in Phase 1 every content change is written to *both* engines.
- **`.os` suffix (the "tag")** — marks a new-engine index. New-engine index names **end in `.os`** (e.g. `cluster_08abc3.working_20260406.os`); old-engine names do not. Visible everywhere: cluster, DB, REST. (NOT an `os::` prefix — that's an obsolete model.)

## Environment

```bash
docker compose -f docker/docker-compose-examples/os-migration/docker-compose.yml up -d
```

| Service | URL | Used for |
|---|---|---|
| Old engine (ES 7.10) | http://localhost:9200 | Direct REST queries |
| Kibana | http://localhost:5601 | Inspect/query the OLD index |
| New engine (OS 3.x) | http://localhost:9201 | Direct REST queries |
| OpenSearch Dashboards | http://localhost:5602 | Inspect/query the NEW index |
| dotCMS | http://localhost:8082 | Admin UI + REST API (`admin:admin`) |

Migration settings live in `dotmarketing-config.properties` (`FEATURE_FLAG_OPEN_SEARCH_PHASE`, `OS_ENDPOINTS=http://localhost:9201`, `OS_AUTH_TYPE=BASIC`, `OS_AUTH_BASIC_USER=admin`, `OS_AUTH_BASIC_PASSWORD=admin`, `OS_TLS_ENABLED=false`).

## Helpers (referenced by the steps below)

**H1 — Find your real index names.** Names contain a timestamp (differ per machine). **New-engine indexes end in `.os`**; old-engine names do not.
```bash
curl -s -u admin:admin http://localhost:8082/api/v1/esindex | jq .   # what dotCMS knows (keeps .os, drops cluster prefix)
curl -s -u admin:admin http://localhost:9200/_cat/indices?v          # OLD engine physical names (no .os)
curl -s -u admin:admin http://localhost:9201/_cat/indices?v          # NEW engine physical names (end in .os)
```
Or **Admin → System → Index**.

**H2 — Which phase did the system end up in?**
- **UI / index list:** Admin → System → Index (or H1). Only old-engine indexes (no `.os`) → Phase 0. New-engine indexes present (`.os`) → Phase 1.
- **Log:** find the migration-phase line (similar to `Migration Phase: ...`). If the shutdown fired you also see the error + "phase reset to 0" lines (see H5).

**H3 — Check the database** (SQL console; new-engine marker is the **`.os` suffix**, not an `os::` prefix):
```sql
SELECT * FROM indicies WHERE index_name LIKE '%.os';      -- NEW engine rows
SELECT * FROM indicies WHERE index_name NOT LIKE '%.os';  -- OLD engine rows
SELECT CASE WHEN index_name LIKE '%.os' THEN 'NEW' ELSE 'OLD' END AS engine, COUNT(*) FROM indicies GROUP BY 1;
```

**H4 — Count documents in an index** (use full physical names from H1; new-engine one ends in `.os`):
```bash
curl -s -u admin:admin http://localhost:9200/<old-index-name>/_count      # e.g. cluster_X.working_20230101
curl -s -u admin:admin http://localhost:9201/<new-index-name.os>/_count    # e.g. cluster_X.working_20260406.os
```

**H5 — Exact log lines of an automatic migration shutdown** (Phase 1/2). Three lines, in order — **two `ERROR` + one `WARN`, NO `FATAL`**:
```
ERROR  OpenSearch configuration error — halting OS migration, dotCMS will fall back to ES-only: <reason>
ERROR  OpenSearch migration halted: invalid configuration detected at startup. Verify OS_ENDPOINTS, OS version, and FEATURE_FLAG_OPEN_SEARCH_PHASE, then restart dotCMS.
WARN   Migration phase reset to PHASE_0_MIGRATION_NOT_STARTED (was PHASE_1_DUAL_WRITE_ES_READS). This change is runtime-only — persist it in dotmarketing-config.properties to survive a restart.
```
`<reason>` is one of:
- unreachable / bad URL: `OpenSearch cluster is not reachable: <cause>. Check OS_ENDPOINTS configuration.`
- wrong version: `OpenSearch version mismatch: expected 3.x but connected cluster reports version <X>. Check OS_ENDPOINTS configuration.`
- same address: `ES and OS clients point to the same endpoint(s): <set>. … Set OS_ENDPOINTS to a separate OpenSearch instance.`

On a successful start you instead see `INFO OS version check passed: <version>`.

## Test Cases

### TC-001 — OpenSearch is down at startup → dotCMS starts normally and keeps serving from ES
- **Objective:** When the migration is turned on (Phase 1) but OpenSearch is unreachable, prove that **dotCMS starts up normally and keeps working using the old engine (ES)** — it must not stop, crash, or hang. The migration switches itself off (drops to Phase 0) and an error is logged.
- **Risk:** High
- **Preconditions:** Old engine running; new engine **stopped** (`docker compose ... up -d elasticsearch kibana`). `FEATURE_FLAG_OPEN_SEARCH_PHASE=1`, `OS_ENDPOINTS=http://localhost:9201`.
- **Steps:**
  1. Start dotCMS. Watch the log until startup finishes — do not kill it even if it pauses trying to reach the new engine.
  2. Check the resulting phase with **H2** (UI and log).
  3. List the indexes with **H1**.
  4. Run **H3** to see whether any new-engine rows (names ending in `.os`) exist.
  5. In the admin UI, create a content item, publish it, then search for it.
- **Expected Result (pass = all true):**
  - **dotCMS starts and runs normally — it does NOT stop, crash, or hang.** Search keeps working against ES.
  - Migration left **off** (Phase 0); log shows the expected error lines (H5: two `ERROR` + one `WARN`, **not** `FATAL`), reason `OpenSearch cluster is not reachable: …. Check OS_ENDPOINTS configuration.`
  - `/api/v1/esindex` (H1) and the UI list **only old-engine indexes** — no new-engine index.
  - DB query (H3) returns **0 rows** ending in `.os`.
  - Create / publish / search all work normally.
- **Replaces:** Scenario B-7 of the legacy `OPENSEARCH_MIGRATION_TEST_PLAN.md` (its old expected result is wrong).

### TC-003 — Migration off (Phase 0) with the new engine down → nothing is affected (control)
- **Objective:** When the migration is **not** active, a down new engine is irrelevant: dotCMS never tries to use it.
- **Risk:** High
- **Preconditions:** Old engine running; new engine **stopped** (`docker compose stop opensearch`). `FEATURE_FLAG_OPEN_SEARCH_PHASE=0` (or no flag).
- **Steps:**
  1. Start dotCMS and watch the startup log.
  2. Create a content item, publish it, search for it.
  3. In Kibana, confirm the new document is in the old-engine index.
- **Expected Result:**
  - dotCMS starts normally. **None** of the H5 shutdown lines appear; no "OS version check passed" line either.
  - Log shows **no attempt** to connect to the new engine.
  - Create / publish / search succeed; document visible in Kibana.

### TC-004 — Turning the migration on must NOT trigger a needless full rebuild
- **Objective:** Switching Phase 0→1, the new engine's index starts empty. dotCMS must **not** mistake "new index empty" for "all indexes empty" and rebuild ES. Existing content is preserved and searchable.
- **Risk:** High
- **Preconditions:** dotCMS running in **Phase 0**.
- **Steps:**
  1. In Phase 0, create and publish at least 5 content items.
  2. Find the old-engine working index name (**H1**) and confirm count > 0 (**H4**).
  3. Stop dotCMS. Set `FEATURE_FLAG_OPEN_SEARCH_PHASE=1`. Restart.
  4. Watch the startup log for any rebuild/reindex messages.
  5. Find the new working index name (**H1**) and check its count (**H4**).
  6. Search the content UI for the items from step 1.
- **Expected Result:**
  - dotCMS starts in Phase 1; a new-engine index appears (newer timestamp).
  - New-engine index count is **0** (back-fill is not part of startup).
  - Old-engine index unchanged, count still > 0.
  - Log does **NOT** contain reindex progress (e.g. no "Reindexation is starting").
  - All pre-existing items still **fully searchable**.

### TC-005 — Creating many indexes at once produces unique, valid names
- **Objective:** Under concurrent load, timestamp-based index names must never collide or come out malformed. 20 simultaneous creates → 20 distinct, well-formed indexes.
- **Risk:** Medium
- **Preconditions:** dotCMS running in **Phase 0**.
- **Steps:**
  1. Fire 20 create requests in parallel:
     ```bash
     seq 20 | xargs -P 20 -I{} curl -s -o /tmp/idx_{}.json -w "%{http_code}\n" \
       -u admin:admin -X PUT http://localhost:8082/api/v1/esindex/create/shards/1
     ```
  2. Collect the 20 returned names from `/tmp/idx_*.json`.
  3. List engine indexes: `curl -s -u admin:admin http://localhost:9200/_cat/indices?v`.
- **Expected Result:**
  - All 20 requests return HTTP **200**.
  - All 20 names are well-formed (no spaces, no empty/null segments, no truncated timestamps).
  - All 20 names are **unique**.
  - The engine shows 20 new indexes with distinct names.

## Acceptance Criteria
- [ ] TC-001: dotCMS starts & serves from ES; migration left at Phase 0; H5 error lines (2 ERROR + 1 WARN, reason "not reachable"); 0 `.os` rows.
- [ ] TC-003: Phase 0 with OS down — no shutdown lines, no OS connection attempt; create/publish/search work.
- [ ] TC-004: Phase 0→1 with existing ES content does NOT trigger a full rebuild; pre-existing content stays searchable.
- [ ] TC-005: 20 concurrent creates → 20 unique, well-formed index names, all HTTP 200.

## Notes
- All cases are Manual. The "automatic migration shutdown" only disables the *migration*; dotCMS itself keeps running on ES.

# [QA-G1] OpenSearch Migration — PR #35632 Acceptance Results (TC-001 → TC-005)

**Issue:** #35635 · **PR under test:** #35632 (Startup Hardening)
**Image tested:** `dotcms/dotcms:opensearch-migration-test-jun10`
**Date:** 2026-06-17

## TL;DR

| TC | Scenario | Result |
|----|----------|:------:|
| TC-001 | Phase 1, OpenSearch unreachable | ✅ PASS |
| TC-002 | Phase 1, wrong OpenSearch version (≠ 3.x) | ✅ PASS |
| TC-003 | Phase 0, OpenSearch down (negative) | ✅ PASS |
| TC-004 | Phase 0 → Phase 1, empty OS index, no false reindex | ✅ PASS |
| TC-005 | 20 concurrent index creations | ⛔ **BLOCKED** — pre-existing bug in `ESIndexResource.createIndex` (details below) |

**4/5 passed.** TC-005 is blocked by a defect in the `createIndex` REST endpoint that is **independent of** the `threadSafeTimestampFormatter` change this PR introduces — the endpoint 500s before the formatter is ever reached, so the test cannot validate the actual fix as written.

Two non-blocking notes for the dev:
- The startup halt is logged at **`ERROR`**, but the acceptance text says **FATAL** — see [Note A](#note-a--log-level-error-vs-fatal).
- `hasEmptyIndices()` correctly avoids a false reindex on Phase 0→1 (TC-004) — confirmed working.

## Test environment

Single-node docker stack (adapted from the `single-core` example compose):

- **Postgres** (`pgvector/pgvector:pg18`)
- **OpenSearch 1.3.20** (`opensearchproject/opensearch:1`) — acts as the ES-compatible primary backend (`DOT_ES_ENDPOINTS=https://opensearch:9200`)
- **OpenSearch 3.7.0** (`opensearchproject/opensearch:3`, security disabled, host `:9201`) — migration target, only used for TC-004
- **dotCMS** (`opensearch-migration-test-jun10`)

The migration phase + OS connection are driven entirely by env vars (no code/topology change needed except adding the 3.x target for TC-004):

```yaml
DOT_FEATURE_FLAG_OPEN_SEARCH_PHASE: '0' | '1'      # phase toggle
DOT_OS_ENDPOINTS:        'http://opensearch3:9200' # per-test
DOT_OS_AUTH_TYPE:        'BASIC'
DOT_OS_AUTH_BASIC_USER:  'admin'
DOT_OS_AUTH_BASIC_PASSWORD: '****'
DOT_OS_TLS_ENABLED:      'false'
```

---

## TC-001 — `haltMigration()` when OpenSearch is unreachable in Phase 1 ✅ PASS

**Setup:** `DOT_FEATURE_FLAG_OPEN_SEARCH_PHASE=1`, `DOT_OS_ENDPOINTS=http://os-dead-endpoint:9201` (no such host), primary ES backend up. Restart dotCMS.

**Expected:** Migration halts, phase resets to Phase 0, FATAL log, no `os::` rows in `indicies`, dotCMS continues on ES.

**Observed (logs):**
```
IndexStartupValidator.validateOSVersion(IndexStartupValidator.java:97)
Caused by: java.net.UnknownHostException: os-dead-endpoint: Name or service not known
ERROR business.ContentletIndexAPIImpl - OpenSearch migration halted: invalid configuration
      detected at startup. Verify OS_ENDPOINTS, OS version, and FEATURE_FLAG_OPEN_SEARCH_PHASE, then restart dotCMS.
WARN  index.IndexConfigHelper$MigrationPhase - Migration phase reset to PHASE_0_MIGRATION_NOT_STARTED
      (was PHASE_1_DUAL_WRITE_ES_READS). This change is runtime-only — persist it in
      dotmarketing-config.properties to survive a restart.
```

**DB (`indicies` table):**
```
 cluster_dotcms-production.live_20260617220609    | live
 cluster_dotcms-production.working_20260617220609 | working
(0 rows with an os:: prefix)
```

**Verdict:** ✅ PASS. Connectivity failure correctly promoted to a hard halt; phase rolled back to `PHASE_0_MIGRATION_NOT_STARTED`; `indicies` contains only ES indices; dotCMS recovered and served on ES (HTTP 200). The `(was PHASE_1_DUAL_WRITE_ES_READS)` text also confirms the phase flag was genuinely read as `1`.

> See [Note A](#note-a--log-level-error-vs-fatal): halt logged at `ERROR`, acceptance text says "FATAL".

---

## TC-002 — `haltMigration()` on version mismatch ✅ PASS

**Setup:** `DOT_FEATURE_FLAG_OPEN_SEARCH_PHASE=1`, `DOT_OS_ENDPOINTS=https://opensearch:9200` (the **reachable** OpenSearch **1.3.20** backend), `DOT_OS_TLS_ENABLED=true`. Restart dotCMS.

**Expected:** Version check detects ≠ 3.x and halts; phase reset; only ES indices registered.

**Observed (logs):**
```
ERROR index.IndexStartupValidator - OpenSearch configuration error — halting OS migration,
      dotCMS will fall back to ES-only: OpenSearch version mismatch: expected 3.x but connected
      cluster reports version 1.3.20. Check OS_ENDPOINTS configuration.
com.dotmarketing.exception.DotRuntimeException: OpenSearch version mismatch: expected 3.x but
      connected cluster reports version 1.3.20.
      at com.dotcms.content.index.IndexStartupValidator.validateOSVersion(IndexStartupValidator.java:101)
ERROR business.ContentletIndexAPIImpl - OpenSearch migration halted: invalid configuration detected at startup.
WARN  index.IndexConfigHelper$MigrationPhase - Migration phase reset to PHASE_0_MIGRATION_NOT_STARTED
      (was PHASE_1_DUAL_WRITE_ES_READS).
```

**DB:** `os_rows=0, total=2` (only ES indices).

**Verdict:** ✅ PASS. This is a genuine **version** check, not a connectivity failure — the TLS connection to the 1.x cluster succeeded and the validator correctly read `1.3.20` and rejected it. Halt + phase reset behaved as in TC-001.

> See [Note A](#note-a--log-level-error-vs-fatal).

---

## TC-003 — Phase 0 startup with OpenSearch unreachable (negative) ✅ PASS

**Setup:** `DOT_FEATURE_FLAG_OPEN_SEARCH_PHASE=0` (default), OS endpoint unreachable. Normal startup.

**Expected:** Clean startup, no FATAL, no OS connection attempted, content searchable on ES.

**Observed:**
- No `haltMigration`, no FATAL/ERROR, no phase reset, no connection attempt/error to the OS endpoint in logs.
- OpenSearch client only *configured* (lazy), never connected:
  ```
  INFO opensearch.ConfigurableOpenSearchProvider - OpenSearch client initialized successfully with endpoints: [...]
  ```
  (config-time only — no connection error despite the endpoint being unreachable)
- Content search returns live data:
  ```
  POST /api/content/_search {"query":"+contenttype:Blog"} → HTTP 200, contentlets returned
  ```
- `indicies`: 2 ES rows, 0 `os::` rows.

**Verdict:** ✅ PASS. In Phase 0 the OS endpoint is never contacted; startup is clean and ES search works.

---

## TC-004 — Phase 0 → Phase 1 with empty OS index, no false reindex ✅ PASS

**Setup:**
1. Phase 0 baseline with existing ES content (starter import → `live`=680 docs, `working`=681 docs).
2. Bring up healthy **OpenSearch 3.7.0** target.
3. Restart dotCMS with `DOT_FEATURE_FLAG_OPEN_SEARCH_PHASE=1`, `DOT_OS_ENDPOINTS=http://opensearch3:9200`.

**Expected:** Version check passes; **no full reindex** ("Reindexation is starting" must NOT appear); new empty OS index created; pre-Phase-1 ES content remains searchable.

**Observed:**
- **No halt, no phase reset** (version check passed against 3.7.0; stayed in Phase 1).
- **No `"Reindexation is starting"`** log — the empty OS index did **not** falsely trigger a full ES reindex. ✅ (core of the `hasEmptyIndices()` phase-aware fix)
- Empty OS indices created on the 3.x target:
  ```
  INFO opensearch.OSIndexAPIImpl - createIndex : Index created: cluster_dotcms-production.working_20260617220609.os with shards: 1
  INFO opensearch.OSIndexAPIImpl - createIndex : Index created: cluster_dotcms-production.live_20260617220609.os    with shards: 1
  ```
- OpenSearch 3.x `_cat/indices` — new indices present and **empty**:
  ```
  cluster_dotcms-production.live_20260617220609.os       0 docs  green
  cluster_dotcms-production.working_20260617220609.os    0 docs  green
  ```
- ES backend **unchanged** (live=680, working=681) and content still searchable in Phase 1 (HTTP 200, results returned).
- `indicies` table now registers all four (dual-write):
  ```
  cluster_dotcms-production.live_20260617220609        | live
  cluster_dotcms-production.live_20260617220609.os     | live
  cluster_dotcms-production.working_20260617220609     | working
  cluster_dotcms-production.working_20260617220609.os  | working
  ```

**Verdict:** ✅ PASS. Phase 0→1 transition with an empty OS index creates the empty OS indices without forcing a reindex; existing ES content is preserved and searchable.

---

## TC-005 — Concurrent index creation (20 parallel requests) ⛔ BLOCKED

**Goal:** Send 20 concurrent `PUT /api/v1/esindex/create/...` requests and verify 20 unique, well-formed index names, validating the `threadSafeTimestampFormatter` thread-safety fix.

**Result:** Could not be executed. **Every** request to the endpoint returns **HTTP 500** before the formatter under test is ever reached. Root cause is a defect in the endpoint itself (not the formatter, not the test environment).

### Root cause

`dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java`:

```java
261  @Deprecated
262  @PUT
263  @Path("/create/{params:.*}")
264  @Produces("text/plain")
265  public Response createIndex(@Context HttpServletRequest httpServletRequest,
266          @Context final HttpServletResponse httpServletResponse,
267          @PathParam("params") String params) {                 // ← path captured into `params`
268      try {
269          InitDataObject init = auth(httpServletRequest, httpServletResponse);   // ← BUG: `params` NOT passed
270          int shards = Integer.parseInt(init.getParamsMap().get("shards"));      // ← get("shards") == null → throws
271          ...
272          indexName = ContentletIndexAPIImpl.threadSafeTimestampFormatter.format(LocalDateTime.now()); // never reached
```
*(line numbers from `main`; the `jun10` test image throws at the identical `:269`)*

The method captures the URL into `@PathParam("params") String params`, but then calls the **2-argument** `auth(...)` overload, which discards it:

```java
120  protected InitDataObject auth(final HttpServletRequest request,
121                                final HttpServletResponse response) {
122      return auth(request, response, null);     // ← null forwarded instead of `params`
123  }
```

`null` flows into `WebResource`, where the param map is built only from the passed string with **no fallback to the request URI**:

```java
// WebResource.java
930  public Map<String,String> paramsMap(){
931      Map<String,String> map = buildParamsMap(UtilMethods.isSet(this.params) ? this.params : BLANK);
                                              // this.params == null → buildParamsMap(BLANK) → empty map
```

So `init.getParamsMap()` is always empty → `getParamsMap().get("shards")` is `null` → `Integer.parseInt(null)` throws `NumberFormatException` → caught and returned as HTTP 500. The captured `params` variable is effectively **dead code**.

### Evidence (live, reproduced)

Server log on every call:
```
ERROR index.ESIndexResource - Error on createIndex. URI: /api/v1/esindex/create/shards/1
java.lang.NumberFormatException: Cannot parse null string
    at com.dotcms.rest.api.v1.index.ESIndexResource.createIndex(ESIndexResource.java:269)
```

Tried 8 input variants (auth `admin@dotcms.com:admin`, `https://localhost:8443`), all identical HTTP 500:

| Request | Result |
|---|---|
| `PUT /api/v1/esindex/create/shards/1` | 500 |
| `PUT /api/v1/esindex/create/shards/1/live/false` | 500 |
| `PUT /api/v1/esindex/create/shards=1/live=false` | 500 |
| `PUT /api/v1/esindex/create/index=x/shards=1` | 500 |
| `PUT /api/v1/esindex/create/shards/1/index/x` | 500 |
| `PUT /api/v1/esindex/create/shards/1?shards=1` (query) | 500 |
| `PUT /api/v1/esindex/create/shards/1` `--data shards=1` (form) | 500 |
| `PUT /api/v1/esindex/create` (no params) | 404 |

Because the param map has no URI fallback, there is **no client-side workaround** — the endpoint cannot be exercised at all.

### Suggested fix (one line)

```java
267      InitDataObject init = auth(httpServletRequest, httpServletResponse, params);
//                                                                          ^^^^^^ forward the captured path
```

This routes `params` into `.params(params)` so `getParamsMap()` is populated and `shards` parses. After this fix (and an image rebuild), TC-005 can run and will finally exercise `threadSafeTimestampFormatter` (line 274).

### Notes for the dev

- This is a **separate, pre-existing bug** in `createIndex`, not a regression from the `threadSafeTimestampFormatter` change. The thread-safety fix may well be correct — we simply cannot reach it via this REST path.
- Same 2-arg `auth(request, response)` pattern is used by sibling endpoints (`clearIndex`, `activateIndex`, `deactivateIndex`). They read params via `indexHelper.getIndexNameOrAlias(init.getParamsMap(), ...)`, which tolerates an empty map (returns null) rather than throwing — so they fail *silently/softly* instead of 500ing. Worth auditing whether those endpoints actually receive their params too.
- Consider adding input validation: a missing/invalid `shards` should return **HTTP 400** with a clear message rather than 500 `NumberFormatException`.
- The endpoint is `@Deprecated`. If TC-005 is meant to target a current API, confirm which endpoint QA should use.

---

## Note A — Log level: `ERROR` vs "FATAL"

The acceptance criteria for TC-001/TC-002 state a **FATAL** log (`"Fatal index configuration error"`) is emitted. In the build tested, the halt is logged at **`ERROR`** level:

```
ERROR business.ContentletIndexAPIImpl - OpenSearch migration halted: invalid configuration detected at startup. ...
ERROR index.IndexStartupValidator - OpenSearch configuration error — halting OS migration, dotCMS will fall back to ES-only: ...
```

The **behavior** is correct (halt + phase reset). This is only a discrepancy between the implementation's log level/wording and the acceptance text. Decide whether to: (a) bump these to `FATAL`, (b) align the acceptance wording to `ERROR`, or (c) leave as-is. Flagging so log-based alerting/assertions key off the right level.

---

## Reproduction summary

```bash
# Baseline / TC-003 (Phase 0)
docker compose up -d db opensearch dotcms

# TC-001 (Phase 1, unreachable OS)
OS_PHASE=1 OS_ENDPOINTS='http://os-dead-endpoint:9201' docker compose up -d dotcms

# TC-002 (Phase 1, wrong version → 1.x backend)
OS_PHASE=1 OS_ENDPOINTS='https://opensearch:9200' OS_TLS=true docker compose up -d dotcms

# TC-004 (Phase 1 → healthy 3.x)
docker compose --profile tc004 up -d opensearch3
OS_PHASE=1 OS_ENDPOINTS='http://opensearch3:9200' OS_TLS=false docker compose up -d dotcms

# Assertions used
docker compose exec -T db psql -U dotcmsdbuser -d dotcms -c \
  "select index_name, index_type from indicies order by index_type;"
curl -sk -u admin:admin   'https://localhost:9200/_cat/indices?v'   # ES backend (1.x)
curl -s                   'http://localhost:9201/_cat/indices?v'    # OS target (3.x)
curl -sk -u admin@dotcms.com:admin -X POST https://localhost:8443/api/content/_search \
  -H 'Content-Type: application/json' -d '{"query":"+contenttype:Blog","limit":1}'
```

--- a/dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java
+++ b/dotCMS/src/main/java/com/dotcms/rest/api/v1/index/ESIndexResource.java
@@ -55,6 +55,7 @@ import com.dotmarketing.portlets.contentlet.model.Contentlet;
 import com.dotmarketing.portlets.contentlet.transform.ContentletToMapTransformer;
 import com.dotmarketing.util.Config;
 import com.dotmarketing.util.Logger;
+import com.dotmarketing.util.UtilMethods;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableMap.Builder;
@@ -264,9 +265,15 @@ public class ESIndexResource {
     @Produces("text/plain")
     public Response createIndex(@Context HttpServletRequest httpServletRequest, @Context final HttpServletResponse httpServletResponse, @PathParam("params") String params) {
         try {
-            InitDataObject init=auth(httpServletRequest, httpServletResponse);
+            InitDataObject init=auth(httpServletRequest, httpServletResponse, params);
 
-            int shards=Integer.parseInt(init.getParamsMap().get("shards"));
+            final String shardsParam = init.getParamsMap().get("shards");
+            if (!UtilMethods.isSet(shardsParam) || !Try.of(() -> Integer.parseInt(shardsParam)).isSuccess()) {
+                Logger.warn(this, "createIndex called with missing/invalid 'shards' param. URI: " + httpServletRequest.getRequestURI());
+                return Response.status(Response.Status.BAD_REQUEST)
+                        .entity("Missing or invalid required 'shards' parameter").build();
+            }
+            final int shards = Integer.parseInt(shardsParam);
             boolean live = init.getParamsMap().containsKey("live") ? Boolean.parseBoolean(init.getParamsMap().get("live")) : false;
             String indexName = init.getParamsMap().get("index");
 
@@ -296,7 +303,7 @@ public class ESIndexResource {
     @Path("/clear/{params:.*}")
     public Response clearIndex(@Context HttpServletRequest httpServletRequest, @Context final HttpServletResponse httpServletResponse, @PathParam("params") String params) throws DotDataException, IOException {
 
-        InitDataObject init=auth(httpServletRequest,httpServletResponse);
+        InitDataObject init=auth(httpServletRequest,httpServletResponse, params);
         String indexName = this.indexHelper.getIndexNameOrAlias(init.getParamsMap(),"index","alias",this.indexAPI);
         return modIndex(httpServletRequest, httpServletResponse, indexName, IndexAction.CLEAR.name());
     }
@@ -404,7 +411,7 @@ public class ESIndexResource {
     @PUT
     @Path("/activate/{params:.*}")
     public Response activateIndex(@Context HttpServletRequest httpServletRequest, @Context final HttpServletResponse httpServletResponse, @PathParam("params") String params) throws DotDataException, IOException {
-        InitDataObject init=auth(httpServletRequest,httpServletResponse);
+        InitDataObject init=auth(httpServletRequest,httpServletResponse, params);
         String indexName = this.indexHelper.getIndexNameOrAlias(init.getParamsMap(),"index","alias",this.indexAPI);
         return modIndex(httpServletRequest, httpServletResponse, indexName, IndexAction.ACTIVATE.name());
     }
@@ -421,7 +428,7 @@ public class ESIndexResource {
     @PUT
     @Path("/deactivate/{params:.*}")
     public Response deactivateIndex(@Context HttpServletRequest httpServletRequest, @Context final HttpServletResponse httpServletResponse, @PathParam("params") String params) throws DotDataException, IOException {
-        InitDataObject init=auth(httpServletRequest,httpServletResponse);
+        InitDataObject init=auth(httpServletRequest,httpServletResponse, params);
         String indexName = this.indexHelper.getIndexNameOrAlias(init.getParamsMap(),"index","alias",this.indexAPI);
         return modIndex(httpServletRequest, httpServletResponse, indexName, IndexAction.DEACTIVATE.name());
     }
@@ -437,7 +444,7 @@ public class ESIndexResource {
     @PUT
     @Path("/close/{params:.*}")
     public Response closeIndex(@Context HttpServletRequest httpServletRequest,@Context final HttpServletResponse httpServletResponse, @PathParam("params") String params) throws DotDataException, IOException {
-        InitDataObject init=auth(httpServletRequest,httpServletResponse);
+        InitDataObject init=auth(httpServletRequest,httpServletResponse, params);
         String indexName = this.indexHelper.getIndexNameOrAlias(init.getParamsMap(),"index","alias",this.indexAPI);
         return modIndex(httpServletRequest, httpServletResponse, indexName, IndexAction.CLOSE.name());
     }
@@ -454,7 +461,7 @@ public class ESIndexResource {
     @Path("/open/{params:.*}")
     public Response openIndex(@Context HttpServletRequest httpServletRequest, @Context final HttpServletResponse httpServletResponse, @PathParam("params") String params) {
         try {
-            InitDataObject init=auth(httpServletRequest, httpServletResponse);
+            InitDataObject init=auth(httpServletRequest, httpServletResponse, params);
             String indexName = this.indexHelper.getIndexNameOrAlias(init.getParamsMap(),"index","alias",this.indexAPI);
             APILocator.getESIndexAPI().openIndex(indexName);
 

